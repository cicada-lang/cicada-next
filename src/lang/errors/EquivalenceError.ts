import { LangError } from "./LangError.js"

export class EquivalenceError extends LangError {
  trace: Array<string> = []

  constructor(public message: string) {
    super(message)
  }
}
