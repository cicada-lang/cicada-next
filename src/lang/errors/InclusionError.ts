import { LangError } from "./LangError.js"

export class InclusionError extends LangError {
  trace: Array<string> = []

  constructor(public message: string) {
    super(message)
  }
}
