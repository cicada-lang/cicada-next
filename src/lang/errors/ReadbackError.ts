import { LangError } from "./LangError.js"

export class ReadbackError extends LangError {
  constructor(public message: string) {
    super(message)
  }
}
