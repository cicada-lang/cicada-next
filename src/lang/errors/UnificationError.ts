import { LangError } from "./LangError.js"

export class UnificationError extends LangError {
  trace: Array<string> = []

  constructor(public message: string) {
    super(message)
  }
}
