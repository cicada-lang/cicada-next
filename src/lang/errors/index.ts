export { ParsingError } from "@cicada-lang/partech"
export * from "./AssertionError.js"
export * from "./ElaborationError.js"
export * from "./EquivalenceError.js"
export * from "./ErrorReport.js"
export * from "./EvaluationError.js"
export * from "./InclusionError.js"
export * from "./InternalError.js"
export * from "./LangError.js"
export * from "./ReadbackError.js"
export * from "./UnificationError.js"
