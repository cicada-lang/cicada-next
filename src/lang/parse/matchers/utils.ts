import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
