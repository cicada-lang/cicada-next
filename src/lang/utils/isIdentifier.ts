const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
