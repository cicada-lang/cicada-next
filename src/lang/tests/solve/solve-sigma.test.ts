import { expect, test } from "vitest"
import * as Errors from "../../errors"
import { expectCodeToFail, runCode } from "../utils"

test("solve Sigma", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  Pair(A, B) = Pair(String, String)
}

`)

  expect(output).toMatchSnapshot()
})

test("solve Sigma -- nested", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  Pair(A, Pair(String, B)) = Pair(String, Pair(String, String))
  Pair(A, Pair(String, String)) = Pair(String, Pair(String, B))
}

`)

  expect(output).toMatchSnapshot()
})

test("solve Sigma -- occur twice", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  Pair(A, Pair(B, B)) = Pair(String, Pair(String, String))
  Pair(A, Pair(B, String)) = Pair(String, Pair(String, B))
}

`)

  expect(output).toMatchSnapshot()
})

test("solve Sigma -- occur in pair", async () => {
  await expectCodeToFail(
    `

solve (x: Type) {
  x = Pair(x, x)
}

`,
    Errors.ElaborationError,
  )
})

test("solve Sigma -- generate const function", async () => {
  const output = await runCode(`

solve (A: Type, B: (x: A) -> Type) {
  exists (x: A) B(x) = exists (_: String) String
}

`)

  expect(output).toMatchSnapshot()
})
