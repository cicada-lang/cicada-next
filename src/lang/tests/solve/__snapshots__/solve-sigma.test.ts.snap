// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve Sigma -- generate const function 1`] = `
"{
  A: String,
  B: (x) => String
}"
`;

exports[`solve Sigma -- nested 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Sigma -- occur twice 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Sigma 1`] = `
"{
  A: String,
  B: String
}"
`;
